administration/src/cards/CardBlueprint.ts

```diff
@@ -5,144 +5,195 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
 import { Region } from '../generated/graphql'
 import { CardConfig } from '../project-configs/getProjectConfig'
 import PlainDate from '../util/PlainDate'
-import RegionExtension from './extensions/RegionExtension'
-import StartDayExtension from './extensions/StartDayExtension'
-import { Extension, ExtensionInstance, JSONExtension, findExtension } from './extensions/extensions'
+import { REGION_EXTENSION_NAME } from './extensions/RegionExtension'
+import Extensions, { Extension, ExtensionKey, ExtensionState, InferExtensionStateType } from './extensions/extensions'
 
 // Due to limited space on the cards
 const MAX_NAME_LENGTH = 30
 // Due to limited space on the qr code
 const MAX_ENCODED_NAME_LENGTH = 50
 
-export type JSONCardBlueprint<E = ExtensionInstance> = {
+export type CardBlueprint = {
   id: number
   fullName: string
-  expirationDate: string | null
-  extensions: (E extends Extension<infer T, unknown> ? JSONExtension<T> : never)[]
+  expirationDate: PlainDate | null
+  extensions: Partial<ExtensionState>
 }
 
-/**
- * Blueprint for a new card. This object contains data about a future card, which will be created.
- */
-export class CardBlueprint {
-  id: number
-  fullName: string
-  expirationDate: PlainDate | null
-  extensions: ExtensionInstance[]
-  cardConfig: CardConfig
-
-  constructor(fullName: string, cardConfig: CardConfig, initParams?: Parameters<CardBlueprint['initialize']>) {
-    this.cardConfig = cardConfig
-    this.fullName = fullName
-    this.expirationDate = initParams ? PlainDate.fromLocalDate(new Date()).add(cardConfig.defaultValidity) : null
-    this.extensions = cardConfig.extensions.map(Extension => new Extension())
-    this.id = Math.floor(Math.random() * 1000000) // Assign some random ID
-    if (initParams) {
-      this.initialize(...initParams)
-    }
+type CardBlueprintKey = keyof Omit<Omit<CardBlueprint, 'extensions'>, 'id'>
+type KeyType = CardBlueprintKey | ExtensionKey | null
+
+const createRandomId = () => Math.floor(Math.random() * 1000000)
```

Comment:
It's quite likely that there will be collisions. if we generate IDs with only 1,000,000 buckets. With 1000 cards, the probability that there is a collision is roughly 40% (see [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem))

What are these ids used for ? Is it possible to just use global counter?

Reply:
Actually not sure, I just kept this from the previous implementation without any logical changes. I was wondering the same though... Not sure what they are used for.

Reply:
Alright, maybe we should create an issue then. Seems quite suspicious.